algorithms/src/polycommit/sonic_pc/data_structures.rs

```diff
@@ -265,6 +284,151 @@ impl<E: PairingEngine> ToBytes for CommitterKey<E> {
 }
 
 impl<E: PairingEngine> CommitterKey<E> {
+    /// Update the committer_key
+    /// The `powers_of_beta_g` and `shifted_powers_of_beta_g` may grow or shrink as per the new supported_degree
+    /// The `enforced_degree_bounds` and `shifted_powers_of_beta_times_gamma_g` will adjust to the new degree_bounds
+    /// The `lagrange_bases_at_beta_g` will optionally adjust to the new degree_bounds
+    /// The `powers_of_beta_times_gamma_g` is only dependent on the `hiding_bound` so doesn't change
+    /// This only works if the SRS max_degree is fixed across specializations
+    /// Note that this implementation is not atomic. If specialize fails halfway through,
+    pub fn update(&mut self, srs: &UniversalParams<E>, degree_info: &DegreeInfo) -> Result<()> {
+        let trim_time = start_timer!(|| "Trimming public parameters");
+
+        // The maximum degree required by the circuit polynomials
+        let supported_degree = degree_info.max_degree;
+        // The maximum degree of the entire SRS
+        let max_degree = srs.max_degree();
+        // The hiding bound for the circuit polynomials
+        let supported_hiding_bound = degree_info.hiding_bound;
+
+        // Update shifted_powers_of_beta_g, shifted_powers_of_beta_times_gamma_g, enforced_degree_bounds
+        match degree_info.degree_bounds.as_ref() {
+            None => {
+                self.shifted_powers_of_beta_g = None;
+                self.shifted_powers_of_beta_times_gamma_g = None;
+                self.enforced_degree_bounds = None;
+            }
+            Some(enforced_degree_bounds) => {
+                let new_highest_degree_bound =
+                    *enforced_degree_bounds.last().ok_or(anyhow!("expecting new degree_bounds"))?;
+                ensure!(new_highest_degree_bound < supported_degree);
+
+                // Retrieve current degree bounds and shifted powers
+                let degree_bounds = self.enforced_degree_bounds.take().unwrap_or_default();
+                let highest_degree_bound =
+                    degree_bounds.iter().copied().sorted().last().map(|b| b + 1).unwrap_or_default();
+                let mut shifted_powers_of_beta_g = self.shifted_powers_of_beta_g.take().unwrap_or_default();
+                let mut shifted_powers_of_beta_times_gamma_g =
+                    self.shifted_powers_of_beta_times_gamma_g.take().unwrap_or_default();
+
+                let shifted_ck_time = start_timer!(|| "Constructing `shifted_powers_of_beta_g`");
+                match new_highest_degree_bound.cmp(&highest_degree_bound) {
+                    Ordering::Greater => {
+                        let lowest_shift_degree = max_degree - new_highest_degree_bound;
+                        let highest_shift_degree = max_degree + 1 - highest_degree_bound;
+                        let mut new_powers = srs.powers_of_beta_g(lowest_shift_degree, highest_shift_degree)?;
+                        new_powers.append(&mut shifted_powers_of_beta_g);
```

Comment:
```suggestion
                        new_powers.extend_from_slice(&shifted_powers_of_beta_g);
```
since we overwrite `shifted_powers_of_beta_g` right afterwards, there is no need to clear that vector

Reply:
My rationale for using `append` was indeed that it moves elements, so I thought this would be faster + of course consumes less memory.

Reply:
as long as the elements are not `Copy`, this is correct :+1:

Reply:
A good point. Just read that moves may be copies under the hood anyway, so here you go: f52cbd381

Reply:
nice; also, a small addition to what I said before: even if the values were not `Copy`, `append` would not consume less memory, because the vector would still retain its allocated space unless `shrink_to_fit` was called afterwards